scico/flax/train/checkpoints.py

```diff
@@ -6,58 +6,77 @@
 # package.
 
 """Utilities for checkpointing Flax models."""
-import os
-from typing import Union
+from pathlib import Path
+from typing import Dict, Union
 
 import jax
 
-try:
-    from tensorflow.io import gfile  # noqa: F401
-except ImportError:
-    have_tf = False
-else:
-    have_tf = True
+import orbax
 
-if have_tf:  # Flax checkpoints require tensorflow
-    from flax.training import checkpoints
+from flax.training import orbax_utils
 
 from .state import TrainState
 
 
-# Flax checkpoints
-def checkpoint_restore(state: TrainState, workdir: Union[str, os.PathLike]) -> TrainState:
+def checkpoint_restore(
+    state: TrainState, workdir: Union[str, Path], ok_no_ckpt: bool = False
+) -> TrainState:
     """Load model and optimiser state.
 
-    Note that naming is slightly different to distinguish from Flax
-    functions.
-
     Args:
         state: Flax train state which includes model and optimiser
             parameters.
-        workdir: checkpoint file or directory of checkpoints to restore
+        workdir: Checkpoint file or directory of checkpoints to restore
             from.
+        ok_no_ckpt: Flag to indicate if a checkpoint is expected. Default:
+                    False, a checkpoint is expected and an error is generated.
 
     Returns:
-        Restored `state` updated from checkpoint file, or if no
-        checkpoint files present, returns the passed-in `state`
-        unchanged.
+        A restored Flax train state updated from checkpoint file is returned.
+        If no checkpoint files are present and checkpoints are not strictly
+        expected it returns the passed-in `state` unchanged.
+
+    Raises:
+        FileNotFoundError: If a checkpoint is expected and is not found.
     """
-    return checkpoints.restore_checkpoint(workdir, state)
+    # Check if workdir is Path or convert to Path
+    workdir_ = workdir
+    if isinstance(workdir_, str):
+        workdir_ = Path(workdir_)
+    if workdir_.exists():
+        orbax_checkpointer = orbax.checkpoint.PyTreeCheckpointer()
+        checkpoint_manager = orbax.checkpoint.CheckpointManager(workdir_, orbax_checkpointer)
+        step = checkpoint_manager.latest_step()
+        if step is not None:
+            target = {"state": state, "config": {}}
+            ckpt = checkpoint_manager.restore(step, items=target)
+            state = ckpt["state"]
+    elif not ok_no_ckpt:
+        raise FileNotFoundError("Could not read from checkpoint: " + workdir)
+
+    return state
 
 
-def checkpoint_save(state: TrainState, workdir: Union[str, os.PathLike]):
-    """Store model and optimiser state.
+def checkpoint_save(state: TrainState, config: Dict, workdir: Union[str, Path]):
+    """Store model, model configuration and optimiser state.
```

Comment:
Store model, model configuration, and optimiser state.

Reply:
Edited.